src/utils/map-state.js

```diff
@@ -1,183 +1,105 @@
-  /** Bearing when current perspective rotate operation started */
-  startBearing: PropTypes.number,
-  /** Pitch when current perspective rotate operation started */
-  startPitch: PropTypes.number,
-  /** Zoom when current zoom operation started */
-  startZoom: PropTypes.number
+  // MAPBOX LIMITS
+  maxZoom: 20,
+  minZoom: 0,
+  maxPitch: 60,
+  minPitch: 0
 };
 
 export default class MapState {
 
   constructor({
     /** Mapbox viewport properties */
+    /** The width of the viewport */
     width,
+    /** The height of the viewport */
     height,
+    /** The latitude at the center of the viewport */
     latitude,
+    /** The longitude at the center of the viewport */
     longitude,
+    /** The tile zoom level of the map. */
     zoom,
-    bearing = 0,
-    pitch = 0,
-    altitude = 1.5,
+    /** The bearing of the viewport in degrees */
+    bearing,
+    /** The pitch of the viewport in degrees */
+    pitch,
+    /**
+    * Specify the altitude of the viewport camera
+    * Unit: map heights, default 1.5
+    * Non-public API, see https://github.com/mapbox/mapbox-gl-js/issues/1137
+    */
+    altitude,
 
     /** Viewport constraints */
-    maxZoom = MAX_ZOOM,
-    minZoom = 0,
-    maxPitch = MAX_PITCH,
-    minPitch = 0,
+    maxZoom,
+    minZoom,
+    maxPitch,
+    minPitch,
 
-    /** Interaction states */
+    /** Interaction states, required to calculate change during transform */
+    /* The point on map being grabbed when the operation first started */
     startPanLngLat,
+    /* Center of the zoom when the operation first started */
     startZoomLngLat,
+    /** Bearing when current perspective rotate operation started */
     startBearing,
+    /** Pitch when current perspective rotate operation started */
     startPitch,
+    /** Zoom when current zoom operation started */
     startZoom
   } = {}) {
-    this.props = {
+    assert(Number.isFinite(width), '`width` must be supplied');
+    assert(Number.isFinite(height), '`height` must be supplied');
+    assert(Number.isFinite(longitude), '`longitude` must be supplied');
+    assert(Number.isFinite(latitude), '`latitude` must be supplied');
+    assert(Number.isFinite(zoom), '`zoom` must be supplied');
+
+    this._state = this._applyConstraints({
       width,
       height,
       latitude,
       longitude,
       zoom,
-      bearing,
-      pitch,
-      altitude,
-      maxZoom,
-      minZoom,
-      maxPitch,
-      minPitch,
+      bearing: Number.isFinite(bearing) ? bearing : defaultState.bearing,
+      pitch: Number.isFinite(pitch) ? pitch : defaultState.pitch,
+      altitude: Number.isFinite(altitude) ? altitude : defaultState.altitude,
+      maxZoom: Number.isFinite(maxZoom) ? maxZoom : defaultState.maxZoom,
+      minZoom: Number.isFinite(minZoom) ? minZoom : defaultState.minZoom,
+      maxPitch: Number.isFinite(maxPitch) ? maxPitch : defaultState.maxPitch,
+      minPitch: Number.isFinite(minPitch) ? minPitch : defaultState.minPitch,
```

Comment:
Nit: Can we extract this logic to a helper above? Like `ensureFinite` that takes the value and a defaultValue. It's terrifying to see so many ternary in a row.